vignettes/sir_models.Rmd

```diff
@@ -199,4 +139,110 @@ matlines(time, t(x[3, , ]), col = cols[["R"]], lty = 1)
 legend("left", lwd = 1, col = cols, legend = names(cols), bty = "n")
 ```
 
+
+## Adding age structure to the model
+
+Adding age structure to the model consists of the following steps, which turn variables into arrays:
+
+- Define the number of age categories as a user parameter `N_age`.
+- Add age structure to each compartment, by adding square brackets to the lvalues.
+- Modify the rvalues to use quantities from the appropriate compartment, by adding an `i` index to the rvalues. These will automatically be looped over.
+- Where an age compartment needs to be reduced into a single compartment/variable, use `sum` (or another array function from https://mrc-ide.github.io/odin/articles/functions.html as appropriate)
+- Define the dimensions of all arrays, for example by setting `dim(S) <- N_age`.
+
+This would simply give `N_age` independent processes equivalent to the first model, scaled by the size of the population in each age category. To actually make this useful, you likely want to add some interaction or transitions between the compartments. An example of this would be to add an age-specific contact matrix, demonstrated below, which defines a different force of infection $\lambda$ for each age group. This is calculated by $$\lambda_i = \frac{\beta}{N} \cdot \sum_{j=1}^{N_{\mathrm{age}}} I_j m_{ij}$$ In the odin code:
+
+```r
+m[, ] <- user() # age-structured contact matrix
+s_ij[, ] <- m[i, j] * I[i]
+lambda[] <- beta / N * sum(s_ij[i, ])
+```
+
+The probability of infection of a susceptible is then indexed by this force of infection:
+
+```r
+p_SI[] <- 1 - exp(-lambda[i] * dt)
+```
+
+Putting this all together, the age structured SIR model is as follows:
+
+
+```{r odin_age, echo = FALSE, results = "asis"}
+r_output(readLines(file.path("sirage.R")))
+```
+
+As before, save the file, and use `odin.dust` to compile the model:
+
+```{r}
+gen_age <- odin.dust::odin_dust("sirage.R")
+```
+
+We can generate an age-structured contact matrix based on the [POLYMOD](https://journals.plos.org/plosmedicine/article?id=10.1371/journal.pmed.0050074) survey by using the [socialmixr](https://github.com/sbfnk/socialmixr) package:
+
+```{r}
+data(polymod, package = "socialmixr")
+
+age.limits = seq(0, 70, 10)
+
+## Get the contact matrix from socialmixr
+contact <- socialmixr::contact_matrix(
+  survey = polymod,
+  countries = "United Kingdom",
+  age.limits = age.limits,
+  symmetric = TRUE)
+
+## Transform the matrix to the (symetrical) transmission matrix
+## rather than the contact matrix
+transmission <- contact$matrix /
+  rep(contact$demography$population, each = ncol(contact$matrix))
```

Comment:
Can you add a brief explanation of what this transformation does in the text above this code block? (I know this is just copied down from my version, but based on how unintuitive it turned out to be I think it'd be useful!)